applications/OptimizationApplication/python_scripts/utilities/opt_line_search.py

```diff
@@ -59,3 +61,71 @@ def GetInfo(self) -> dict:
                 'scaled_step': self.step}
 
         return info
+    
+class BB_Step(object):
+    @classmethod
+    def GetDefaultParameters(cls):
+        return Kratos.Parameters("""{
+            "type"               : "BB_step",
+            "init_step"          : 0,
+            "max_step"           : 0,
+            "gradient_scaling": ": inf_norm"
+        }""")
+
+    def __init__(self, parameters: Kratos.Parameters, optimization_problem: OptimizationProblem):
+        parameters.ValidateAndAssignDefaults(self.GetDefaultParameters())
+        self.init_step = parameters["init_step"].GetDouble()
+        self.max_step = parameters["max_step"].GetDouble()
+        self.__optimization_problem = optimization_problem
+        self.__gradient_scaling = parameters["gradient_scaling"].GetString()
+
+    @time_decorator()
+    def ComputeStep(self) -> float:
+        algorithm_buffered_data = ComponentDataView("algorithm", self.__optimization_problem).GetBufferedData()
+
+        if not algorithm_buffered_data.HasValue("search_direction"):
+            raise RuntimeError(f"Algorithm data does not contain computed \"search_direction\".\nData:\n{algorithm_buffered_data}" )
+
+        if self.__gradient_scaling == "inf_norm":
+            norm = KratosOA.ExpressionUtils.NormInf(algorithm_buffered_data["search_direction"])
+        elif self.__gradient_scaling == "l2_norm":
+            norm = KratosOA.ExpressionUtils.NormL2(algorithm_buffered_data["search_direction"])
+        elif self.__gradient_scaling == "none":
+            norm = 1.0
+        else:
+            raise RuntimeError("\"gradient_scaling\" has unknown type.")
+        
+        if self.__optimization_problem.GetStep() == 0:
+            self.unscaled_step = self.init_step
+        else:
+            current_search_direction = algorithm_buffered_data.GetValue("search_direction", 0)
+            previous_search_direction = algorithm_buffered_data.GetValue("search_direction", 1)
+            y = previous_search_direction - current_search_direction
+            d = algorithm_buffered_data.GetValue("control_field_update", 1)
+            if not math.isclose(KratosOA.ExpressionUtils.InnerProduct(d,y), 0.0, abs_tol=1e-16):
+                self.unscaled_step = KratosOA.ExpressionUtils.InnerProduct(d,d) / KratosOA.ExpressionUtils.InnerProduct(d,y)
+            if math.isclose(KratosOA.ExpressionUtils.InnerProduct(d,y), 0.0, abs_tol=1e-16):
+                self.unscaled_step = self.max_step
+            if math.isclose(KratosOA.ExpressionUtils.InnerProduct(d,d), 0.0, abs_tol=1e-16):
+                self.unscaled_step = 0.0
```

Comment:
the result of the inner product could be stored in a variable to avoid computing it 3 times